### ISSUE TITLE
[Models] BART model output is incorrect shape

### ISSUE THREAD

**Issue by nitishgupta** [allenai/allennlp#4550]
## Checklist

- [x] I have verified that the issue exists against the `master` branch of AllenNLP.
- [x] I have read the relevant section in the [contribution guide](https://github.com/allenai/allennlp/blob/master/CONTRIBUTING.md#bug-fixes-and-new-features) on reporting bugs.
- [x] I have checked the [issues list](https://github.com/allenai/allennlp/issues) for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/allenai/allennlp/pulls) for existing proposed fixes.
- [x] I have checked the [CHANGELOG](https://github.com/allenai/allennlp/blob/master/CHANGELOG.md) and the [commit log](https://github.com/allenai/allennlp/commits/master) to find out if the bug was already fixed in the master branch.
- [x] I have included in the "Description" section below a traceback from any exceptions related to this bug.
- [x] I have included in the "Related issues or possible duplicates" section beloew all related issues and possible duplicate issues (If there are none, check this box anyway).
- [x] I have included in the "Environment" section below the name of the operating system and Python version that I was using when I discovered this bug.
- [x] I have included in the "Environment" section below the output of `pip freeze`.
- [x] I have included in the "Steps to reproduce" section below a minimally reproducible example.

## Description

The output of the BART model is incorrect here -- https://github.com/allenai/allennlp-models/blob/master/allennlp_models/generation/models/bart.py#L206

It should be `(batch_size, seq_len, vocab_size)` but is `(batch_size, 1, vocab_size)`

<details>
<summary><b>Python traceback:</b></summary>
<p>

```
```

</p>
</details>

## Related issues or possible duplicates

- None

## Environment

OS: Linux

Python version: 3.7.7

<details>
<summary><b>Output of <code>pip freeze</code>:</b></summary>
<p>

```
-e git+git@github.com:allenai/allennlp.git@e53d18580807dabff707f618f7e148c98d25da18#egg=allennlp
-e git+git@github.com:allenai/allennlp-models.git@45f85ce67b869e6f33fb0a38bd9ae17bd99f287c#egg=allennlp_models
apex @ git+https://github.com/NVIDIA/apex.git@459de22d59c64e30fd4b368c368c5b74e269f3dd
appdirs==1.4.4
attrs==19.3.0
bert-score==0.3.5
black==19.10b0
bleach==3.1.5
blis==0.4.1
boto3==1.14.29
botocore==1.17.29
catalogue==1.0.0
certifi==2020.6.20
cffi==1.14.1
chardet==3.0.4
click==7.1.2
codecov==2.1.8
colorama==0.4.3
conllu==3.1.1
coverage==5.2.1
cryptography==3.0
cycler==0.10.0
cymem==2.0.3
dateparser==0.7.6
docutils==0.15.2
editdistance==0.5.3
en-core-web-lg @ https://github.com/explosion/spacy-models/releases/download/en_core_web_lg-2.3.1/en_core_web_lg-2.3.1.tar.gz
en-core-web-sm @ https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-2.3.1/en_core_web_sm-2.3.1.tar.gz
filelock==3.0.12
flake8==3.8.3
flaky==3.7.0
ftfy==5.8
future==0.18.2
googledrivedownloader==0.4
h5py==2.10.0
idna==2.10
importlib-metadata==1.7.0
jeepney==0.4.3
Jinja2==2.11.2
jmespath==0.10.0
joblib==0.16.0
jsonnet==0.16.0
jsonpickle==1.4.1
jsons==1.2.0
keyring==21.2.1
kiwisolver==1.2.0
livereload==2.6.2
lunr==0.5.8
lxml==4.5.2
Markdown==3.2.2
markdown-include==0.5.1
MarkupSafe==1.1.1
matplotlib==3.3.0
mccabe==0.6.1
mkdocs==1.1.2
mkdocs-material==5.5.5
mkdocs-material-extensions==1.0
mkl-fft==1.1.0
mkl-random==1.1.1
mkl-service==2.3.0
more-itertools==8.4.0
murmurhash==1.0.2
mypy==0.782
mypy-extensions==0.4.3
nltk==3.5
nr.collections==0.0.1
nr.databind.core==0.0.16
nr.databind.json==0.0.13
nr.interface==0.0.3
nr.metaclass==0.0.5
nr.parsing.date==0.2.0
nr.pylang.utils==0.0.3
nr.stream==0.0.4
nr.utils.re==0.1.0
numpy==1.18.5
olefile==0.46
overrides==3.1.0
packaging==20.4
pandas==1.1.0
pathspec==0.8.0
pathtools==0.1.2
Pillow @ file:///tmp/build/80754af9/pillow_1594307325547/work
pkginfo==1.5.0.1
plac==1.1.3
pluggy==0.13.1
preshed==3.0.2
protobuf==3.12.2
py==1.9.0
py-cpuinfo==7.0.0
py-rouge==1.1
pycodestyle==2.6.0
pycparser==2.20
pydoc-markdown @ git+https://github.com/NiklasRosenstein/pydoc-markdown.git@f0bf8af1db4f11581c19d206d4ed1ab34b4854c1
pyflakes==2.2.0
Pygments==2.6.1
pymdown-extensions==7.1
pyparsing==2.4.7
pytest==5.4.3
pytest-benchmark==3.2.3
pytest-cov==2.10.0
python-dateutil==2.8.1
pytz==2020.1
PyYAML==5.3.1
readme-renderer==26.0
regex==2020.7.14
requests==2.24.0
requests-toolbelt==0.9.1
responses==0.10.15
rfc3986==1.4.0
ruamel.yaml==0.16.10
ruamel.yaml.clib==0.2.0
s3transfer==0.3.3
sacremoses==0.0.43
sacrerouge==0.0.4
scikit-learn==0.23.1
scipy==1.5.2
seaborn==0.10.1
SecretStorage==3.1.2
sentencepiece==0.1.91
six==1.15.0
spacy==2.3.2
srsly==1.0.2
tensorboardX==2.1
thinc==7.4.1
threadpoolctl==2.1.0
tokenizers==0.8.1rc1
toml==0.10.1
torch==1.5.1
torchvision==0.6.0a0+82fd1c8
tornado==6.0.4
tqdm==4.48.0
transformers==3.0.2
twine==3.2.0
typed-ast==1.4.1
typing-extensions==3.7.4.2
typish==1.7.0
tzlocal==2.1
Unidecode==1.1.1
urllib3==1.25.10
wasabi==0.7.1
watchdog==0.10.3
wcwidth==0.2.5
webencodings==0.5.1
word2number==1.1
zipp==3.1.0
```

</p>
</details>

## Steps to reproduce

<details>
<summary><b>Example source:</b></summary>
<p>

```
x = torch.LongTensor([[0, 133, 766, 9]])
m = torch.BoolTensor([[True, True, True, True]])
bart = BartForConditionalGeneration.from_pretrained("facebook/bart-large", output_past=True)
y = bart(input_ids=x, attention_mask=m, decoder_input_ids=x, decoder_attention_mask=m)[0]
y.size()   # output == torch.Size([1, 1, 50265])
```

</p>
</details>

**Comment by nitishgupta** [comment]
@dirkgr - I saw that you recently added BART to the allennlp-models, hence you might be the best person to know about this.

**Comment by nitishgupta** [comment]
Apologies for the false alarm -- I found a solution here: https://github.com/huggingface/transformers/issues/6353
One line fix; `use_cache=True` needs to be passed to the BART model's `forward`.

I tried to test the CNN/DM BART model here but, (1) I first ran into an issue with the data link in the config, (2) I downloaded the tar manually but the dataset reader was taking forever to untar and I killed the process.

Is it possible to release a small sample dataset to test the model. Once I'm able to test, I can send a PR with the change if that helps.

**Comment by dirkgr** [comment]
I did not dig into the details, but is this something we should change on our end?

What was the issue with the data link in the config?

Untarring will indeed take a long time. It has to create many small files, which is never ideal. I'd be happy to look at various PRs that address this, for example:
 * change the data format so that the data lives in a few large files
 * show a progress bar while un-tarring
 * add a small dataset and make it the default in the config, add the large dataset in a commented-out line in the config

**Comment by nitishgupta** [comment]
Yes, I think a one-line change would be needed. But before I could do that I am struggling with getting past the reader. 

1. When I use the `tar.gz` URL given in the config I get the error:
```
File "/mnt/castor/seas_home/n/nitishg/code/allennlp-models-master/allennlp_models/generation/dataset_readers/cnn_dm.py", line 144, in _read
    "Story with url '%s' and hash '%s' not found" % (url, url_hash)
allennlp.common.checks.ConfigurationError: Story with url '<?xml version='1.0' encoding='UTF-8'?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Details>No such object: allennlp-public-data/cnndm-combined-data-2020.07.13.tar.gz!cnndm-combined-data-2020.07.13/url_lists/all_train.txt</Details></Error>' and hash '27438822fec35dec31e16c6b893b266d47633d92' not found
```

2. When I download the `tar` locally and run I get a 
```
FileNotFoundError: file /scratch1/nitishg/cnndm-combined-data-2020.07.13.tar.gz!cnndm-combined-data-2020.07.13/url_lists/all_train.txt not found
```

I am running this on OpenSuse linux. Would you know what might be the issue.

**Comment by nitishgupta** [comment]
Weirdly I do not get an error when I try to train this model, but get one doing something very similar. 
I'm closing this now, will reopen if I face issues.

**Comment by dirkgr** [comment]
Alright, but I am quite interested in the issue with the data link? We try pretty hard to make it so you can run the training configs out of the box (as long as the data isn't copyright protected, at least).

**Comment by nitishgupta** [comment]
Okay, yeah I just confirmed and trying to train using the URL provided does run into an issue:
```
  File "/mnt/castor/seas_home/n/nitishg/code/allennlp-models-master/allennlp_models/generation/dataset_readers/cnn_dm.py", line 144, in _read
    "Story with url '%s' and hash '%s' not found" % (url, url_hash)
allennlp.common.checks.ConfigurationError: Story with url 'http://web.archive.org/web/20070716092219id_/http://us.cnn.com:80/2007/US/07/13/btsc.obrien.criminallyinsane/index.html' and hash 'ee8871b15c50d0db17b0179a6d2beab35065f1e9' not found
```

This runs if I download the `.tar.gz` locally and point to that. Haven't dug into what the issue might be.

P.S. I'm using OpenSuse linux.

**Comment by dirkgr** [comment]
That sounds like it's a case of #4501. You probably tried to run the config, and killed the extraction half-way through. Now it thinks the extraction is already there, even though only have the files are in it. You can fix this by removing the extraction directory, and then running the config again. You'll have to be patient until the extraction is complete.

I'll close this, but feel free to re-open if we didn't catch all the issues.

**Comment by leonleeldc** [comment]
Hey, when I run bart model, I got errors, assert len(layer_cache) == 4 in decoder_cache_dict = self._decoder_cache_to_dict(decoder_cache). Not sure what wrong? May you give help there?

**Comment by epwalsh** [comment]
What version of transformers are you using?

**Comment by leonleeldc** [comment]
>>> import transformers
>>> transformers.__version__
'4.0.1'

**Comment by epwalsh** [comment]
You need at least 4.1: https://github.com/allenai/allennlp/blob/main/setup.py#L68

**Comment by leonleeldc** [comment]
Oh, I see. Let me update my transformers then. Thanks a lot!

**Comment by leonleeldc** [comment]
Hey, the problem for Bart is resolved. But I have a new problem in deploying DDP. When the data size is as high as 50m (the input is a pair of sentences (10 tokens each)), the CPU memory will be a problem. I am using V100 gpu workers with 8 gpu nodes (16G for each). The cpu memory is 480G. I am deploying an DNN model. Even if I only use 4 gpu nodes with batch size as small as 256. After running some time, the CPU memory will be exhausted. Not sure how I can fix this. I noticed allennlp 2.0 has a new dataset reader, which does not need to load all data into memory. Do you think whether I should upgrade to 2.0 or whether there is a solution to current 1.3 version? Thanks,

**Comment by leonleeldc** [comment]
I am trying to use allennlp 2.0. But I found that there are quite minor changes. For example, token becomes token_class, Tensorboard becomes TensorBoard, dataloaders becomes data_loaders. For those changes, I have found and replaced them. But for the following, not sure what I should do. 
And I didn't find where similarity_function is now. 
    from allennlp.modules.similarity_functions import SimilarityFunction
ModuleNotFoundError: No module named 'allennlp.modules.similarity_functions'
And so on. May you give us help on how to handle such changes.

**Comment by epwalsh** [comment]
I'm not sure what `SimilarityFunction` was exactly. What version of AllenNLP were you using? Sounds like you might be looking for [Attention](https://github.com/allenai/allennlp/blob/main/allennlp/modules/attention/attention.py) or [MatrixAttention](https://github.com/allenai/allennlp/blob/main/allennlp/modules/matrix_attention/matrix_attention.py)?

**Comment by leonleeldc** [comment]
I had a check and found that SimilarityFunction is in fact from allennlp_0.9. I just copied them into new version (v1.3). I copied them into allennlp_2.0. Looks no problem for it. I am still trying to make 2.0 work though.

**Comment by dirkgr** [comment]
Similarity functions were already removed in AllenNLP 1.0. I don't know why. I'm afraid that was even before my time. Maybe @matt-gardner remembers?

**Comment by matt-gardner** [comment]
`SimilarityFunction` was an attempt to let you experiment with different similarity functions when computing attentions and similar operations.  The trouble was that it was implemented very inefficiently.  We ended up pushing each of the `SimilarityFunction` implementations directly to subclasses of `Attention` and `MatrixAttention`, where they could be done much more efficiently.  Same functionality in all cases, just with different abstractions and more efficient implementations.
